lib/naxum/src/error_handling.rs

```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+#[allow(unused_macros)]
+macro_rules! impl_service {
+    ( $($ty:ident),* $(,)? ) => {
+        impl<S, R, F, Res, Fut, $($ty,)*> Service<R>
+            for HandleError<S, F, ($($ty,)*)>
+        where
+            S: Service<R> + Clone + Send + 'static,
+            S::Response: IntoResponse + Send,
+            S::Error: Send,
+            S::Future: Send,
+            F: FnOnce($($ty),*, S::Error) -> Fut + Clone + Send + 'static,
+            Fut: Future<Output = Res> + Send,
+            Res: IntoResponse,
+            $( $ty: FromMessageHead<()> + Send,)*
+            R: MessageHead + Send + 'static,
+        {
+            type Response = Response;
+            type Error = Infallible;
+
+            type Future = future::HandleErrorFuture;
+
+            fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+                Poll::Ready(Ok(()))
+            }
+
+            #[allow(non_snake_case)]
+            fn call(&mut self, req: R) -> Self::Future {
+                let f = self.f.clone();
+
+                let clone = self.inner.clone();
+                let inner = std::mem::replace(&mut self.inner, clone);
+
+                let future = Box::pin(async move {
+                    let (mut parts, body) = req.into_parts();
+
+                    $(
+                        let $ty = match $ty::from_message_head(&mut parts, &()).await {
+                            Ok(value) => value,
+                            Err(rejection) => return Ok(rejection.into_response()),
+                        };
+                    )*
+
+                    let req = match R::from_parts(parts, body) {
+                        Ok(value) => value,
+                        Err(rejection) => return Ok(rejection.into_response()),
+                    };
+
+                    match inner.oneshot(req).await {
+                        Ok(res) => Ok(res.into_response()),
+                        Err(err) => Ok(f($($ty),*, err).await.into_response()),
+                    }
+                });
+
+                future::HandleErrorFuture { future }
+            }
+        }
+    }
+}
+
+impl_service!(T1);
+impl_service!(T1, T2);
+impl_service!(T1, T2, T3);
+impl_service!(T1, T2, T3, T4);
+impl_service!(T1, T2, T3, T4, T5);
+impl_service!(T1, T2, T3, T4, T5, T6);
+impl_service!(T1, T2, T3, T4, T5, T6, T7);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15);
+impl_service!(T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16);
```

Comment:
What the hell is this

Reply:
Haha, so much fun! That's how axum supports handler functions or extractor functions that take no args, 1 arg, 2 args...they all need to be explicitly created! I used the same approach here which does lead to some funky code